wooey/models/favorite.py

```diff
@@ -0,0 +1,18 @@
+from django.db import models
+from django.conf import settings
+from ..django_compat import GenericForeignKey
+from django.contrib.contenttypes.models import ContentType
+
+class Favorite(models.Model):
+
+    user = models.ForeignKey(settings.AUTH_USER_MODEL, related_name='favorites', blank=True, null=True)
+
+    content_type = models.ForeignKey(ContentType)
+    object_id = models.PositiveIntegerField()
+    content_object = GenericForeignKey('content_type', 'object_id')
```

Comment:
Just for Scripts and Files I imagine (can't think of anything else that would make sense). The scrapbook and the script favourites are the same thing.

You're right about the convoluted way to get the objects. At the moment the list of `Favorite` objects is got matching the right content type (id/etc.) and used to generate a list of object IDs. This list of IDs is then queried against the database as a list of IDs using `__in` (so 2 queries; plus another for the ContentType if it's not in memory).

It's been a while but you're right that with GFKs you lose prefetches (there is no real link in the ORM). The way around it I've used before is to use an adapter table, e.g. `ScriptFavorites` which would link from the `Script` to `Favorite`. This contains just the `script_id` and the `favorite_id` — the favorite info remains in a `Favorite` object.

The difference is that this allows you to traverse the ORM.